src/server/pkg/storage/storage-format-spec.md

```diff
@@ -0,0 +1,84 @@
+  - Index entries index into other tar streams. 
+  - Content entries are standard tar entries of type file or directory.
+- All tar streams are sorted with respect to the lexicographical order of the header names. Duplicate header names are allowed and are interpreted as a merge (described in the merge section).
+- This storage format:
+  - requires an implementation of a content-addressed chunk storage layer. 
+  - uses protocol buffers as the serialization format for indexes.
+
+**Chunk Storage Layer:**
+
+- The chunk storage layer should:
+  - be able to convert arbitrary byte streams into content-addressed chunks and store them.
+  - provide an interface for creating and accessing ranges within/across these byte streams.
+- A reference to data within a chunk is stored as a DataRef protocol buffer message which includes the chunk's content-address(hash), offset, and size.
+- Referencing a range across multiple chunks requires stringing together multiple of these single chunk references. 
+- The hash of the data may also be included if it does not match the chunk hash (the data is only part of the chunk).
+
+```
+message DataRef {
+	Chunk chunk
+	string hash
+	int64 offset_bytes
+	int64 size_bytes
+}
+
+message Chunk {
+	string hash
+}
+```
+
+**Index Entry:**
+
+- An index entry contains an index into other tar streams. 
+- The header name of an index entry is the same as the first header in the indexed tar stream. 
+- Index entries may index a tar stream with only one tar entry or a range of tar entries.
+    - Index entries that index one tar entry have headers with type flag set to 'i' and have additional metadata about the single tar entry. 
+    - Index entries that index a range of tar entries have headers with type flag set to 'r' and have additional metadata about the range of tar entries. 
+- The index is stored within an Index protocol buffer message that is serialized into the content section of the index entry. 
+
+```
+message Index {
+	Range Range
+	DataOp dataOp
+}
+
+message Range {
+	string lastPath
+	// other stuff (bloom filter, stats, etc.)
+}
+
+message DataOp {
+	repeated DataRef dataRef
+	Op op
+	repeated Tag tags
+}
+
+enum Op {
+	APPEND
+	OVERWRITE
+	DELETE
+}
+
+message Tag {
+	int64 id
+	int64 size_bytes
+}
+```
+
+- The indexed tar stream is represented by a data operation which may contain a sequence of data references that are associated with the operation (there should be no data references for a delete). The sequence of data references should be stored in the order they should be read to get the final state of the underlying tar stream.
+- A sequence of tags can be associated with a data operation to communicate the relative ordering of data between data operations during merge. The tags are ordered in correspondence with the data, such that reading the data references and tags together provides enough information to determine the mapping from tag to data.
```

Comment:
Tags are the only part of this that I still don't feel like I understand. I think maybe just an example of what a valid `DataOp` with tags would look and mean, and an explanation of what it means without the tags would be helpful.

Reply:
I guess it would just be easier to have a picture:

          dr1       dr2          dr3
        [........][.............][............] data references for a file
[................][.............][........................] byte stream (brackets identify chunk split points)
        [...][.......][.....][.........][.....] tags for a file (would be datums in output file)
         t1    t2     t3      t4      t5

The tags provide enough information to know where we need to insert some new data that belongs in the middle of the file without having to store all the information that a DataRef requires. You would interpret this by collating the tags with the resulting byte stream from the data references.

In terms of DataOps, I was thinking we could have one per tar entry and they would become relevant when merging tar entries with the same header name (path).

Reply:
Ok I think I get it, is it true that the sum of the `size_bytes` fields in the tags should be the same as the sum of the `size_bytes` fields in the data refs? In the picture it looks like the byte stream is bigger, which suggests a stream where only some of the data is tagged, which seems weird to me.

Reply:
Yes, that is true. The byte stream has multiple files, the data refs and tags shown apply to one of the files in the byte stream.